src/System.Net.Sockets/src/System/Net/Sockets/SocketAsyncContext.Linux.cs

```diff
@@ -0,0 +1,1200 @@
+                {
+                    if (errno != Interop.Error.EAGAIN && errno != Interop.Error.EWOULDBLOCK)
+                    {
+                        errorCode = SocketPal.GetSocketErrorForErrorCode(errno);
+                        return true;
+                    }
+
+                    errorCode = SocketError.Success;
+                    return false;
+                }
+
+                bytesSent += sent;
+
+                bool isComplete = sent == 0 ||
+                    (buffer != null && count == 0) ||
+                    (buffers.IsInitialized && bufferIndex == buffers.Count);
+                if (isComplete)
+                {
+                    errorCode = SocketError.Success;
+                    return true;
+                }
+            }
+        }
+
+        private static void QueueCompletion(AsyncOperation operation)
+        {
+            ThreadPool.QueueUserWorkItem(o => ((AsyncOperation)o).Complete(), operation);
+        }
+
+        public unsafe void HandleEvents(uint events)
+        {
+            lock (_closeLock)
+            {
+                if ((events & Interop.libc.EPOLLERR) != 0)
+                {
+                    int errno;
+                    uint optLen = (uint)sizeof(int);
+                    int err = Interop.libc.getsockopt(_fileDescriptor, Interop.libc.SOL_SOCKET, Interop.libc.SO_ERROR, &errno, &optLen);
+                    if (err == -1)
+                    {
+                        // TODO: throw an appropiate exception
+                        throw new Exception(string.Format("HandleEvents getsockopt: {0}", Interop.Sys.GetLastError()));
+                    }
+
+                    // TODO: error handling
+                }
+
+                // TODO: rethink ordering of HUP/RDHUP w.r.t. IN/OUT.
+
+                if ((events & Interop.libc.EPOLLHUP) != 0)
+                {
+                    // Drain queues and unregister fd
+
+                    Queue<AcceptOrConnectOperation> acceptOrConnectQueue;
+                    Queue<TransferOperation> sendQueue;
+                    Queue<TransferOperation> receiveQueue;
+                    lock (_queueLock)
+                    {
+                        acceptOrConnectQueue = _acceptOrConnectQueue.Stop();
+                        sendQueue = _sendQueue.Stop();
+                        receiveQueue = _receiveQueue.Stop();
+
+                        if (_registeredEvents != 0)
+                        {
+                            Interop.Error errorCode;
+                            if (!_engine.TryUnregister(ref _handle, _fileDescriptor, 0, out errorCode))
+                            {
+                                if (errorCode != Interop.Error.EBADF)
+                                {
+                                    // TODO: throw an appropiate exception
+                                    throw new Exception(string.Format("HandleEvents HUP: {0}", errorCode));
+                                }
```

Comment:
We don't want to free this handle in the case of an exception?  Is that because the caller might try to recover and continue using it?